### PR TITLE
WIP: Preprocessing revitalized

### PR DESCRIPTION
**Pull request by wasade**
First pass. This has at its core a different approach to processing. The central idea is to not tie to a compute resource which greatly reduces the amount of code necessary to manage compute. Instead, the model is that the notebooks will assume they have full access to hammer on the machine they're executing on. This way, we can rely on making direct system calls (i.e., `!print_qiime_config.py`) as well as QIIME's parallelism within a single host through just forking multiple processes. On a compute cluster, the expectation is that an individual would request an entire node for a reasonable amount of walltime to do the work.

- [x] setup `runipys` similar to [IAB](https://github.com/gregcaporaso/An-Introduction-To-Applied-Bioinformatics/blob/master/runipynbs.py)
- [x] tie into `.travis.yml`, and set `export AG_TESTING=True`
- [x] resolve filtering notebook 
- [x] resolve OTU picking notebook
- [x] prepare for meta analyses (merging tables, mappings, etc) 
- [x] resolve diversity analysis notebook
- [x] resolve taxonomy summaries notebook
- [x] resolve generating of participant results notebook
- [ ] resolve project wide summary

### CODE REVIEW

**Comment by wasade** [comment]
...also, this encourages much smaller notebooks which is refreshing to say the least.

File: americangut/util.py [hunk 1]
```diff
@@ -20,6 +22,87 @@
 __email__ = "mcdonadt@colorado.edu"
```
Comment (mortonjt) [inline americangut/util.py]: Do you still want to keep this email address?

File: americangut/util.py [hunk 2]
```diff
@@ -178,15 +261,29 @@ def fetch_metadata_xml(accession):
...
+
+    description = sample.find('DESCRIPTION')
+    metadata['Description'] = description.text.strip('" ')
+
     return metadata
 
 
-def fetch_study(accession, metadata_path, fasta_path):
+def fetch_study(accession):
     """Fetch and dump a full study
```
Comment (mortonjt) [inline americangut/util.py]: It'll be preferable to have a Parameters section here - up to you.

File: americangut/util.py [hunk 3]
```diff
@@ -201,9 +298,10 @@ def fetch_study(accession, metadata_path, fasta_path):
         md = fetch_metadata_xml(sample)
         all_md[qiimedb_samplename] = md
         all_cols.update(md)
-
+        break
         # write out fasta
         try:
+            raise ValueError()
```
Comment (mortonjt) [inline americangut/util.py]: Since this is raising a ValueError, is the following couple of lines really necessary?  It looks like they are being short-circuited

Comment (wasade) [inline americangut/util.py]: thanks for catching. ya, intentionally broken right now.

File: americangut/util.py (hunk 2 above)
Comment (wasade) [inline americangut/util.py]: this is old code, so i agree that refreshes on it are needed but not the immediate focus.

File: americangut/util.py (hunk 1 above)
Comment (wasade) [inline americangut/util.py]: meh

File: ipynb/filter_sequences_for_blooms.ipynb [hunk 4]
```diff
@@ -0,0 +1,246 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Early in the American Gut Project, it was observed that some organisms bloomed likely as a result of increased shipping time and delay between when samples were collected and when they were put on ice (more detail can be found [here](http://americangut.org/?page_id=277)). The purpose of this notebook is to apply the filter developed in order to bioinformatically subtract these observed bloom sequences from fecal samples. It is important to apply this filter when combining data with the American Gut as to remove a potential study-effect bias as all fecal data in the American Gut has had this filter applied. The specific steps covered are:\n",
```
Comment (mortonjt) [inline ipynb/filter_sequences_for_blooms.ipynb]: This is a bit nit-picky.

But the term "apply the filter" is a little ambiguous at first, since the "filter" has not been identified.

Is it possible to reword it to something like
"The purpose of this notebook is to illustrate how these organisms are identified and filtered out in our bioinformatics software pipeline."

File: ipynb/pick_otus.ipynb [hunk 5]
```diff
@@ -0,0 +1,228 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Pick OTUs (for full and trimmed data) at approximately genus level resolution (97% similarity) using SortMeRNA, closed reference against Greengenes 13_8."
```
Comment (mortonjt) [inline ipynb/pick_otus.ipynb]: Just to sanity check.

Can we assume that people looking at these notebooks have prior knowledge about OTU picking?

If not, it may be desirable to at least include a link.

File: ipynb/pick_otus.ipynb [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "filtered_sequences       = agu.get_existing_path(agenv.paths['filtered-sequences'])\n",
+    "filtered_sequences_100nt = agu.get_existing_path(agenv.paths['filtered-sequences-100nt'])\n",
```
Comment (mortonjt) [inline ipynb/pick_otus.ipynb]: Do we want to mention what are trimmed sequences and how the trimming was done?

File: ipynb/pick_otus.ipynb [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+      "    main()\n",
+      "  File \"/Users/mcdonadt/miniconda3/envs/agdev/bin/pick_closed_reference_otus.py\", line 224, in main\n",
+      "    status_update_callback=status_update_callback)\n",
+      "  File \"/Users/mcdonadt/miniconda3/envs/agdev/lib/python2.7/site-packages/qiime/workflow/upstream.py\", line 506, in run_pick_closed_reference_otus\n",
+      "    close_logger_on_success=close_logger_on_success)\n",
+      "  File \"/Users/mcdonadt/miniconda3/envs/agdev/lib/python2.7/site-packages/qiime/workflow/util.py\", line 122, in call_commands_serially\n",
+      "    raise WorkflowError(msg)\n",
+      "qiime.workflow.util.WorkflowError: \n",
+      "\n",
+      "*** ERROR RAISED DURING STEP: Make OTU table\n",
```
Comment (mortonjt) [inline ipynb/pick_otus.ipynb]: Did this fail?  If so, will this need to be corrected?

**Comment by mortonjt** [comment]
I think this is mostly okay.  There are a few comments in pick_otus.

File: americangut/util.py [hunk 8]
```diff
@@ -20,6 +22,87 @@
...
+    """
+    path = get_path(path)
+
+    if not os.path.exists(path):
+        raise IOError('%s does not exist.' % path)
+
+    return path
+
+
 def pick_rarifaction_level(id_, lookups):
```
Comment (gregcaporaso) [inline americangut/util.py]: spelling error, should be *rarefaction*

**Comment by gregcaporaso** [comment]
*Remove the recruited bloom sequences from the demultiplexed sequence data*

Why do this at the sequence level and not the OTU level? (And if this isn't the kind of feedback you're looking for now, just let me know.)

**Comment by rob-knight** [comment]
Because doing it at the OTU level was too broad in earlier tests, though perhaps this should be revisited.

**Comment by gregcaporaso** [comment]
*We also need to specify what specific metadata category and value correspond indicate what samples are fecal.* extra words here

**Comment by gregcaporaso** [comment]
The bloom sequences are used as reference sequences for 97% closed-reference OTU picking, and then all sequences that cluster into the OTUs defined by those centroids are removed. So the "range" of sequences that are removed is the same size as if you did this at the OTU level, but the centroids are specifically defined by the bloom sequences. This is interesting and makes sense, but I think it'd be good to describe this specifically (where you mention how sequences are recruited above) as it wasn't the way that I would have thought to do this.

**Comment by gregcaporaso** [comment]
One really high-level comment on the notebooks - if these were written in markdown instead of ipynb they'd be easier to diff, have others contribute to, etc. This is what we're doing with IAB now (see [my MSL blog post](https://www.mozillascience.org/an-introduction-to-applied-bioinformatics-at-mozsprint) about this).

**Comment by wasade** [comment]
That is awesome. Yes, will look at that right now

**Comment by gregcaporaso** [comment]
[This video](https://www.youtube.com/watch?v=s9-aZrX5CY8) shows how readers can submit changes as PRs 100% through their web browser.

**Comment by gregcaporaso** [comment]
Reviewed ``get_sequences_and_metadata.ipynb``, that all makes sense.

**Comment by gregcaporaso** [comment]
In ``pick_otus.ipynb``, you shouldn't need to pass ``-r`` or ``-t`` anymore - they're defaulting to the files you're wanting to use (I think).

**Comment by gregcaporaso** [comment]
It could help with debugging if your two ``pick_closed_reference_otus.py`` commands were in different cells.

**Comment by gregcaporaso** [comment]
@wasade, added some comments on your notebooks. These generally look good to me (clearly *a lot* better than the massive notebook you linked me to). If you want me to review a few more of these as you get them together let me know.

**Comment by wasade** [comment]
Not fully, I want to be able to exercise these notebooks via travis so I
need to replace the reference used as GG 97% isn't feasible for testing due
to 15min indexing time with SortMeRNA (and not able to incl. in repo due to
size)

**Comment by wasade** [comment]
Thank you @gregcaporaso, this is awesome. I'm going to finish doing the indirection to get the testing up for pick OTUs, tie into .travis, and then do the ipymd bit you linked.

**Comment by gregcaporaso** [comment]
Ah, right, that makes sense (though you could probably add a .qiime_config
to travis if you wanted to, but no difference either way).

File: americangut/notebook_environment.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    'gg-otus-100nt': 'otus/gg-13_8-97-percent-otus-with-100nt',
+    'gg-otus-biom': 'otus/gg-13_8-97-percent-otus/otu_table.biom',
+    'gg-otus-100nt-biom':
+        'otus/gg-13_8-97-percent-otus-with-100nt/otu_table.biom',
+}
+
+
+def _assert_environment():
+    if qiime.__version__ != '1.9.1':
+        raise ImportError("QIIME 1.9.1 is not in the environment.")
```
Comment (squirrelo) [inline americangut/notebook_environment.py]: Maybe put current version info in the error?

Comment (wasade) [inline americangut/notebook_environment.py]: :+1:

File: ipynb/primary-processing/3-pick_otus.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+```
+
+Before we go too far, let's make sure the files we need are present.
+
+```python
+>>> filtered_sequences       = agu.get_existing_path(agenv.paths['filtered-sequences'])
+>>> filtered_sequences_100nt = agu.get_existing_path(agenv.paths['filtered-sequences-100nt'])
+```
+
+And, let's make sure that the output files we need do not already exist.
```
Comment (squirrelo) [inline ipynb/primary-processing/3-pick_otus.md]: output files we create?

**Comment by squirrelo** [comment]
Couple small things.

File: ipynb/index.ipynb [hunk 11]
```diff
@@ -4,12 +4,20 @@
...
     "\n",
-    "* Primary processing pipeline [Notebook](module2_v1.0.ipynb)\n",
-    "* Filter bloom sequences [Notebook](FilterAndPickOTUs.ipynb)\n",
+    "1. [Get sequences and metadata](primary-processing/1-get_sequences_and_metadata.ipynb)\n",
+    "2. [Filter sequences for blooms](primary-processing/2-filter_sequences_for_blooms.ipynb)\n",
+    "3. [Pick OTUs](primary-processing/3-pick_otus.ipynb)\n",
+    "4. [Prepare metaanalyses](primary-processing/4-prepare_metaanalyses.ipynb)\n",
+    "5. [Alpha diversity calculations](primary-processing/5-alpha_diversity.ipynb)\n",
+    "6. [Beta diversity calculations](primary-processing/6-beta_diversity.ipynb)\n",
+    "7. [Taxonomy summaries](primary-processing/7-taxonomy_summaries.ipynb)\n",
```
Comment (wasade) [inline ipynb/index.ipynb]: needs 7.5

File: americangut/notebook_environment.py [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+
+    if find_executable('print_qiime_config.py') is None:
+        raise EnvironmentError("The QIIME executables are not in $PATH.")
+
+    if find_executable('mod2_pcoa.py') is None:
+        raise EnvironmentError("The AG scripts are not in $PATH.")
+_assert_environment()
```
Comment (squirrelo) [inline americangut/notebook_environment.py]: Is this supposed to be here?

Comment (wasade) [inline americangut/notebook_environment.py]: yes. if the environment is not sane, the bail as early as possible

File: americangut/parallel.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+        A file-like object to write a list of successful sample IDs too
+    fail_fp : file-like object
+        A file-like object to write a list of unsuccessful sample IDs too,
+        and any associated error information
+    partitions : iterable
+        Yields a function and an iterable of IDs. It is expected that this
+        function will return a dict of {str: list}.
+    """
+    pool = mp.Pool(processes=mp.cpu_count())
+
```
Comment (mortonjt) [inline americangut/parallel.py]: Would it be worthwhile to pass in the number of cores as a parameter?

That why the user can specify exactly how many cores that they would like to use.  This could also help prevent users from accidentally over-swamping their system.

Comment (wasade) [inline americangut/parallel.py]: It introduces either a config file or yet another env var.

**Comment by wasade** [comment]
In the interest of simplifying review, as this PR is titanic, I'm going to close it out and open individual PRs for each notebook one-by-one.
